### PR TITLE
Linux: Installing/Updating packages with uppercase versions

### PR DESCRIPTION
**Pull request by edvilme**
Addresses #39105 #24419 #41844, https://github.com/NuGet/Home/issues/13844, etc

When attempting to install or update a tool on Linux, the system looks for the file with the version as is, instead of lowercasing it. Packages are ideally stored with lowercase versions on them according to the NuGet team, so the system fails to find them.

Instead, versions are lowercased and the `VersionFolderPathResolver` API is used for path operations to be compliant with changes from the NuGet codebase.

### CODE REVIEW

File: src/Cli/dotnet/ToolPackage/ToolPackageDownloader.cs [hunk 1]
```diff
@@ -295,7 +295,7 @@ private static async Task<NuGetVersion> DownloadAndExtractPackage(
             }
 
             // Extract the package
-            var nupkgDir = Path.Combine(packagesRootPath, packageId.ToString(), version.ToString());
+            var nupkgDir = Path.Combine(packagesRootPath, packageId.ToString(), version.ToNormalizedString().ToLowerInvariant());
```
Comment (nkolev92) [inline src/Cli/dotnet/ToolPackage/ToolPackageDownloader.cs:298]: I think using APIs like VersionFolderPathResolver might be better than manually doing these calculations. 
It minimizes the number of places changes would need to be made should something change.

Unfortunately, I don't really know what other consequence that might have throughout the code, so that idea would require more investigation.,

Comment (edvilme) [reply src/Cli/dotnet/ToolPackage/ToolPackageDownloader.cs:298]: For a lot of them, I think they are just string modifications so I would agree it would be better to use the API. I can try replacing it where it seems relevant. Thanks!!

File: src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs [hunk 2]
```diff
@@ -218,7 +214,10 @@ public async Task<string> GetPackageUrl(PackageId packageId,
             SourceRepository repository = GetSourceRepository(source);
             if (repository.PackageSource.IsLocal)
             {
-                return Path.Combine(repository.PackageSource.Source, $"{packageId}.{resolvedPackageVersion}.nupkg");
+                return Path.Combine(
```
Comment (nkolev92) [inline src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:217]: This is probably feedback for a future PR, but using the FindPackagesByIdResource handles all of the logic in this method for you. 
It abstracts whether you're talking to a public or private feed, and it can help you download

File: src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs [hunk 3]
```diff
@@ -307,9 +306,7 @@ await GetPackageMetadataAsync(packageId.ToString(), packageVersion, packagesSour
             return (source, packageVersion);
         }
 
-        private string GetNupkgUrl(string baseUri, PackageId id, NuGetVersion version) =>
-            baseUri + id.ToString() + "/" + version.ToNormalizedString() + "/" + id.ToString() +
-            "." + version.ToNormalizedString() + ".nupkg";
+        private string GetNupkgUrl(string baseUri, PackageId id, NuGetVersion version) => new VersionFolderPathResolver(baseUri).GetPackageFilePath(id.ToString(), version);
```
Comment (nkolev92) [inline src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:309]: The nupkgs are in the folders for id/version, so this will miss out on that and generate an incorrect link. 

Per: https://[api.nuget.org/v3/index.json](https://api.nuget.org/v3/index.json), 
```json
     {
            "@id": "https://api.nuget.org/v3-flatcontainer/",
            "@type": "PackageBaseAddress/3.0.0",
            "comment": "Base URL of where NuGet packages are stored, in the format https://api.nuget.org/v3-flatcontainer/{id-lower}/{version-lower}/{id-lower}.{version-lower}.nupkg"
        },```
        
Eventually FindPackagesByIdResource could be used to just "solve all of that for you.

Comment (nkolev92) [reply src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:309]: You probably want a test for this.

**Review by nkolev92 — COMMENTED** [review]
I think having an E2E or at least some unit tests for the installation path generated would help long term.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs [hunk 4]
```diff
@@ -1443,7 +1443,7 @@ private void WritePackageDependenciesDesignTime()
                 foreach (var package in _lockFile.Libraries)
                 {
                     var packageVersion = package.Version.ToNormalizedString();
-                    string packageId = $"{package.Name}/{packageVersion}";
+                    string packageId = $"{package.Name}/{packageVersion.ToLowerInvariant()}";
```
Comment (nagilson) [inline src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:1446]: @nkolev92 Does the lockFile.Libraries folder hold the same assumption as the global packages folder which is that all of the versions will be lowercased? 

This applies to the ResolvePackageDependencies file as well.

File: src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs (hunk 2 above)
Comment (nagilson) [reply src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:217]: @edvilme, may you create an issue on the SDK to update some of our nuget api calls? Then we can decide how to prioritize that work in a different PR and you can resolve this comment.

File: src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs (hunk 3 above)
Comment (nagilson) [reply src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:309]: Thank you @nkolev92  for taking a look at these changes and noticing this. 

It looks like `FindPackagesByIdResource` is not available yet: https://github.com/search?q=org%3ANuGet%20FindPackagesByIdResource%20%20&type=code

Is this correct? In this case, I suppose we would just use the old code but call the to lower case normalization function on the version string? Or we could call GetPackageFilePath twice and edit it a bit to match the existing definition?

File: src/Cli/dotnet/ToolPackage/ToolPackageDownloader.cs [hunk 5]
```diff
@@ -107,11 +107,12 @@ public IToolPackage InstallPackage(PackageLocation packageLocation, PackageId pa
                     }
                     NuGetVersion packageVersion = nugetPackageDownloader.GetBestPackageVersionAsync(packageId, versionRange, packageSourceLocation).GetAwaiter().GetResult();
 
-                    rollbackDirectory = isGlobalTool ? toolDownloadDir.Value: Path.Combine(toolDownloadDir.Value, packageId.ToString(), packageVersion.ToString());
+                    rollbackDirectory = isGlobalTool ? toolDownloadDir.Value: new VersionFolderPathResolver(toolDownloadDir.Value).GetInstallPath(packageId.ToString(), packageVersion);
 
                     if (isGlobalTool)
                     {
-                        NuGetv3LocalRepository nugetPackageRootDirectory = new(Path.Combine(_toolPackageStore.GetRootPackageDirectory(packageId).ToString().Trim('"'), packageVersion.ToString()));
+                        // TODO: NuGetv3LocalRepository already has a VersionFolderPathResolver member, so this can probably be simplified
```
Comment (nagilson) [inline src/Cli/dotnet/ToolPackage/ToolPackageDownloader.cs:114]: If you don't plan on including this in this PR (which you dont have to) I would suggest you remove the TODO and add a remark in the issue mentioned in an earlier comment.

File: src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs [hunk 6]
```diff
@@ -131,14 +131,10 @@ public async Task<string> DownloadPackageAsync(PackageId packageId,
...
             }
 
-            string nupkgPath = downloadFolder == null || !downloadFolder.HasValue
-                ? Path.Combine(_packageInstallDir.Value, packageId.ToString(),
-                    resolvedPackageVersion.ToNormalizedString(),
-                    $"{packageId}.{resolvedPackageVersion.ToNormalizedString()}.nupkg")
-                : Path.Combine(downloadFolder.Value.Value,
-                    $"{packageId}.{resolvedPackageVersion.ToNormalizedString()}.nupkg");
-
+            var pathResolver = new VersionFolderPathResolver(downloadFolder == null || !downloadFolder.HasValue ? _packageInstallDir.Value : downloadFolder.Value.Value);
```
Comment (nagilson) [inline src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:134]: Ah this code reads so much better and more easily than before, thank you 😄

File: src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs [hunk 7]
```diff
@@ -6,6 +6,7 @@
 using Microsoft.DotNet.Tools;
 using Microsoft.Extensions.EnvironmentAbstractions;
 using NuGet.Frameworks;
+using NuGet.Packaging;
```
Comment (nagilson) [inline src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs:9]: On `ToolRestoreCommand` line 157, I would lowercase the package.Version going into the Tool Resolver cache. I would also do this in the localToolsResolverCache line 26, 149. I think its possible that data could come to bite us later if we access it somewhere without lowercasing the version.

Also going to ping @dsplaisted to see if he has any thoughts.

File: src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs [hunk 8]
```diff
@@ -6,6 +6,7 @@
...
 using NuGet.ProjectModel;
 using NuGet.Versioning;
```
Comment (nagilson) [inline src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs:11]: @dsplaisted Do you think we should lowercase the tool version on the tool manifests going forward? I don't think we need to do this or necessarily should, because the tool can still have different version casing; its just that it gets treated the same under the nuget package folder.

**Review by nagilson — COMMENTED** [review]
There's still some pending feedback I'd like to resolve / hear from others on, but great job on your first PR in the SDK! 😊 This was a lot more challenging than expected which is our fault so I do apologize for that. I think you did a great job independently investigating and contacting the nuget team so great work 👏

File: src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs (hunk 8 above)
Comment (edvilme) [reply src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs:11]: I can see the argument for both sides. I tried to avoid making more changes than needed. However, to me it would make a lot of sense to have them lowercase _everywhere_ to avoid confusion with users, and to potentially avoid issues like this from ever happening. 
Yet again, I also understand why a user might expect to see their version _as is_ (e.g., Camel Case, or casing having actual meaning on their workflow) - although I think this would be less common

File: src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs (hunk 2 above)
Comment (edvilme) [reply src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:217]: https://github.com/dotnet/sdk/issues/44169

File: src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs (hunk 3 above)
Comment (nkolev92) [reply src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:309]: It's available: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Protocol/Resources/FindPackageByIdResource.cs. 

You need to do `await sourceRepository.GetResourceAsync<FindPackagesByIdResource>()`

Comment (nkolev92) [reply src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:309]: > Is this correct? In this case, I suppose we would just use the old code but call the to lower case normalization function on the version string?

Yeah, you can do either, probably easier to read if you're explicit with the old code.
Especially if you're going to refactor to the FindPackagesByIdResource later.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs (hunk 4 above)
Comment (nkolev92) [reply src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:1446]: I don't think the libraries lowercases them.
It uses the true casing from nuspec afaik. 
The expectation is that the comparison is case insensitive there. 

I actually think that within the assets file, the id of the package will always have the consistent case.

File: src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs (hunk 8 above)
Comment (nagilson) [reply src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs:11]: Well put and I agree. I lean towards not lowercasing it for now but let's see what his thoughts are.

File: src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs (hunk 3 above)
Comment (edvilme) [reply src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:309]: Ok, I will revert to the previous code and annotate on the issue for refactoring later, thanks!

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs (hunk 4 above)
Comment (edvilme) [reply src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:1446]: Is it ok then if I don't lowercase them then?

Comment (nkolev92) [reply src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:1446]: If you check the usage of packageId, it's used in allowItemSpecs.Contains which is an OrdinalIgnoreCase hashset.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs [hunk 9]
```diff
@@ -1508,7 +1508,7 @@ HashSet<string> GetPackageDependencies()
                     return results;
                 }
 
-                static string GetPackageId(LockFileTargetLibrary package) => $"{package.Name}/{package.Version.ToNormalizedString()}";
+                static string GetPackageId(LockFileTargetLibrary package) => $"{package.Name}/{package.Version.ToNormalizedString().ToLowerInvariant()}";
```
Comment (nkolev92) [inline src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:1511]: Same thing probably applies here. 
The checks using this string are ordinal ignore case.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs (hunk 4 above)
Comment (nagilson) [reply src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:1446]: It should be ok but I would leave it as is as Nikolche suggests ; I think putting this conversation in a comment would be helpful.

File: src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs (hunk 8 above)
Comment (dsplaisted) [reply src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs:11]: If the version in the tool resolver cache would end up being displayed to the user, then I think it would be better to preserve the original case.  If not, then it might be better to lowercase it, but there could be bugs either way,

**Review by dsplaisted — COMMENTED** [review]
LGTM, thanks!

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageDependencies.cs
Comment (edvilme) [inline src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageDependencies.cs:1]: @nagilson @dsplaisted @nkolev92 Should I lowercase them here?

File: src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs (hunk 8 above)
Comment (edvilme) [reply src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs:11]: Should I then preserve the original one? @dsplaisted

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageDependencies.cs
Comment (nkolev92) [reply src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageDependencies.cs:1]: They don't seem to used for matching, but as metadata in items generated by this task. 

True casing is the right approach here, so I don't think you need it.

File: src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs (hunk 8 above)
Comment (edvilme) [reply src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs:11]: I will look to see it these strings (a) get printed out to the user, (b) are used to search for files in disk to make the decision in the meantime :)

Comment (edvilme) [reply src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs:11]: Since they are used to generate paths to files, I advocate towards having them lowercased

Comment (nagilson) [reply src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs:11]: Based on what you said, I think lowercasing them is a good idea; good investigation there :)
I would advocate for testing it around a little bit, especially with the old behavior then going to the new behavior. Example: Installing a tool to create a tool manifest with the older version, then trying to use the tool on the newer branch

Comment (edvilme) [reply src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs:11]: From my testing, it still should work regardless of the casing on the manifest file
image

**Comment by edvilme** [comment]
Hello @nagilson @dsplaisted, 
Please let me know if there is anything else on my end for this PR, thx! :D

File: src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs (hunk 8 above)
Comment (nagilson) [reply src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs:11]: Thank you for taking the time to test that, glad it works 🥳

**Review by nagilson — COMMENTED** [review]
One thing I thought was a side effect of this change at first is
image
that dotnet tool list will only show a lowercase version. This doesnt matter in terms of a breaking change on Linux because even in 7.0 it didnt quite work properly.

However on Windows it did, but it looks like even on Windows it was lowercased in the past, so that's not a concern.

File: src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs (hunk 7 above)
Comment (nagilson) [reply src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs:9]: `ToolRestoreCommand.PackageHasBeenRestored` Loads the Cache of tool restore commands previously executedand then the locations of those tools that were previously restored -> `TryGetMatchingRestoredCommand` -> `LocalTolsResolverCache` `Save` uses `RestoredCommand.Executable` to cache the location of the package, it includes `Version` as well which is not lowercased but that `Version` is not used to generate a path so it's ok, also I think we want to store the true version of the package and not the lowercased one for nuget's directory purposes 
-> `ToolRestoreCommand.InstallPackages` has `IToolPackage` with `Command` with `Executable`  -> `ToolPackageDownloader.InstallPackage` -> Uses Nuget Version in the `Version` but `toolReturnPackageDirectory` for the directory which happens from `GetPackageDirectory` which already lowercased the version.................

TLDR: This is the correct approach and we dont need to lowercase this version I think 😵

**Review by nagilson — APPROVED** [review]
:shipit: Fantastic work, this was much harder than anticipated. You did a great job digging through the code, thank you. I would not expect a typical new grad to be able to do that. Let's get this merged!

**Comment by edvilme** [comment]
> :shipit: Fantastic work, this was much harder than anticipated. You did a great job digging through the code, thank you. I would not expect a typical new grad to be able to do that. Let's get this merged!

Thanks! It was very interesting and useful to get to know the tooling a bit better :)
